### DISCUSSION TITLE
Blocks ending with "end"

### DISCUSSION THREAD

**Discussion by StephanSchmidt** [alecthomas/participle#257] [Q&A]
I want to end blocks with "end". It seems to work with when I have one element to match, but not an array [] where it includes end somehow.

My understanding the default parser in participle is C-oriented, which would mean { } blocks.

Do I need to change the lexer/parser to have

    def x() :
      ...
    end

blocks?

Or do I need to tell blocks somehow (E.g. Body) that "end" is a keyword that should not be included in the match?  Or do I do something wrong?

**Comment by alecthomas** [comment]
I'm not sure what the exact issue is (if you can include some code that compiles and illustrates your problem, that would be helpful), but this kind of thing will work:

```go
type Def struct {
  Name string `"def" @Ident "(" ")" ":"`
  Body []*Stmt `@@* "end"`
}
```

A raw string such as `"end"` will match the token, but not assign it anywhere.

**Comment by StephanSchmidt** [comment]
I have something that you suggested, but he doesn't find the end when there are two functions. I'll take another look to find the bug in my code, thanks.

```Go
type Function struct {
	Name   string     `"def" @Ident`
	Args   []*Arg     `"(" ( @@ ( "," @@ )* )? ")"`
	Return string     `"-" ">" @Ident ":"`
	Body   []*Command `@@* "end"`
}
```

**Comment by StephanSchmidt** [comment]
It seems to come from this

```Go
type Command struct {
	Pos   lexer.Position
	Print *Print `( @@ `
	Expression *Expression `| @@ )`
}
```

With only one, `Print`, it works. With two it can't find "end"

**Comment by StephanSchmidt** [comment]
I can't seem to make it work. I've created a repo

https://github.com/StephanSchmidt/hiro
